_posts/2023-04-21-extensibility-roadmap.markdown

```diff
@@ -0,0 +1,154 @@
+
+### Storage plugins become extensions
+
+The storage API in OpenSearch has proven to be quite stable. There may be no need to change these extensions across OpenSearch versions. New features and improvements in the respective remote storage clients (e.g. S3 or Azure) happen out of band from OpenSearch distributions, so these extensions can be upgraded and improved without needing to wait for a new OpenSearch release. We also want partners to maintain their own storage and remove the idea of "first class storage".
+
+With the addition of features like remote-backed indexes and searchable snapshots, the storage plugins (e.g. S3, Azure, GCP, HDFS) are on the critical path for both indexing and search operations. These plugins will not be able to tradeoff any performance penalty as that is likely to make for an unacceptable user experience. We’ll use the in-process support for extensions to move these plugins to the extensions model. We could abstract just reading or writing, or even support multiple versions of Lucene side-by-side, or even different storage engines and initialize one engine per index.
+
+### Replication as an extension
+
+With the introduction of segment replication (segrep) node resources need to be allocated to perform file copy. Today, there are settings to define limits on data transfer rates (for segrep & recovery) to prevent these functions from consuming valuable resources required for indexing & search. Moving this functionality to a separate node local JVM allows us to control maximum resource consumption (cpu/memory), avoid unnecessary throttling, and have no impact on read & write performance. We can therefore define new extension points on the engine to support segrep implementations that can run as a sidecar, and provide the opportunity to plug in an implementation based on storage requirement (remote, node-node). This would be either in process or a separate node local process and be integrated with storage plugins to support remote store as replication source.
+
+### Extensions in other programming languages
+
+With extensions designed to operate remotely we can support out of process extensions written in other languages running on the JVM, and remote extensions written in any other language hosted externally. By enabling polyglot applications, or porting the extension SDKs to other programming languages we would also lower the barrier of entry for authoring extensions in languages such as JRuby. With TypeScript extensions on GraalVM we could support in-proc extensions.
+
+### Extensions in other technologies
+
+An extension can not be entirely implemented in a Lambda, as it must maintain a network connection with OpenSearch and shares some state information. It would be possible to create a thin client for this purpose with Lambda implementation internally to enable Lambda extensions.
```

Comment:
Nit:
```suggestion
An extension can not be entirely implemented in a Lambda, as it must maintain a network connection with OpenSearch and shares some state information. It would be possible to create a lean client for this purpose with Lambda implementation internally to enable Lambda extensions.
```